Source/Connector.swift

```diff
@@ -0,0 +1,167 @@
+
+    init(
+        centralManager: CBCentralManager,
+        delegateWrapper: CBCentralManagerDelegateWrapper
+    ) {
+        self.centralManager = centralManager
+        self.delegateWrapper = delegateWrapper
+    }
+
+    /**
+     Establishes connection with a given `Peripheral`.
+     For more information see `CentralManager.establishConnection(with:options:)`
+    */
+    func establishConnection(with peripheral: Peripheral, options: [String: Any]? = nil)
+        -> Observable<Peripheral> {
+            return .deferred { [weak self] in
+                guard let strongSelf = self else {
+                    return Observable.error(BluetoothError.destroyed)
+                }
+
+                let connectionObservable = strongSelf.createConnectionObservable(for: peripheral, options: options)
+
+                let waitForDisconnectObservable = strongSelf.delegateWrapper.didDisconnectPeripheral
+                    .filter { $0.0 == peripheral.peripheral }
+                    .take(1)
+                    .do(onNext: { [weak self] _ in
+                        guard let strongSelf = self else { return }
+                        strongSelf.disconnectingBox.write { $0.remove(object: peripheral.identifier) }
+                    })
+                    .map { _ in peripheral }
+                let isDisconnectingObservable: Observable<Peripheral> = Observable.create { observer in
+                    var isDiconnecting = strongSelf.disconnectingBox.read { $0.contains(peripheral.identifier) }
+                    let isDisconnected = peripheral.state == .disconnected
+                    // it means that peripheral is already disconnected, but we didn't update disconnecting box
+                    if isDiconnecting && isDisconnected {
+                        strongSelf.disconnectingBox.write { $0.remove(object: peripheral.identifier) }
+                        isDiconnecting = false
+                    }
+                    if !isDiconnecting {
+                        observer.onNext(peripheral)
+                    }
+                    return Disposables.create()
+                }
+                return waitForDisconnectObservable.amb(isDisconnectingObservable)
+                    .flatMap { _ in connectionObservable }
+            }
+    }
+
+    fileprivate func createConnectionObservable(
+        for peripheral: Peripheral,
+        options: [String: Any]? = nil
+    ) -> Observable<Peripheral> {
+        return Observable.create { [weak self] observer in
+            guard let strongSelf = self else {
+                observer.onError(BluetoothError.destroyed)
+                return Disposables.create()
+            }
+
+            guard !peripheral.isConnected else {
+                observer.onError(BluetoothError.peripheralAlreadyConnected(peripheral))
+                return Disposables.create()
+            }
+
+            let connectingStarted = strongSelf.connectingBox.compareAndSet(
```

Comment:
I don't like that there is no global connection resource which is held uniquely. Current flow in theory may allow 2 simultaneously connected users. Let's assume that thread 1 and 2 are at line 96 and device is disconnected and not connecting to any other device. Thread 1 executes line 98 setting connecting state and after a while it goes to 138 by setting connecting to false after successful connection. Then thread 2 starts executing line 98 and is able to call `connect` even though thread 1 already owns a connection.

Reply:
To clarify semantics we want to achieve there are three states which are important:

1) Device is disconnected. Any thread is allowed to take ownership of a resource.
2) Device is connecting/connected. One thread (winner) owns a resource, other threads reject with an error.
3) Device is disconnecting. Only owner from step 1 can transition to this state. Other threads now are allowed to wait.